pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/agent/types"
+	crdv1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/util/ip"
+)
+
+type ruleType int
+
+type ruleEvent uint8
+
+const (
+	unicast   ruleType = 0
+	igmp      ruleType = 1
+	multicast ruleType = 2
+
+	ruleAdd    ruleEvent = 1
+	ruleUpdate ruleEvent = 2
+	ruleDelete ruleEvent = 3
+
+	groupAddressIndex = "groupAddressIndex"
+)
+
+type mcastItem struct {
+	groupAddress net.IPNet
+	ruleIDs      map[string]*uint16
+}
+
+type mcastIGMPRule struct {
+	ruleID       string
+	priority     *uint16
+	groupAddress []string
+	event        ruleEvent
+}
+
+// multicastController handles only IGMP protocol. It stores the group address to RuleID map and
+// also rule to group addresses map
+type multicastController struct {
+	ofClient openflow.Client
+
+	ifaceStore interfacestore.InterfaceStore
+	ruleCache  *ruleCache
+	queue      workqueue.RateLimitingInterface
+	// key is the multicast group address.
+	mcastItemRuleCache cache.Indexer
+
+	// key is ruleID while the value is list of group addresses.
+	ruleIDGroupAddressMap map[string]sets.String
+	ruleIDGroupMapMutex   sync.RWMutex
+}
+
+func groupAddressKey(obj interface{}) (string, error) {
+	group := obj.(*mcastItem)
+	return group.groupAddress.IP.String(), nil
+}
+
+func groupAddressRuleIndexFunc(obj interface{}) ([]string, error) {
+	item := obj.(*mcastItem)
+	ruleIDs := make([]string, len(item.ruleIDs))
+	for ruleID := range item.ruleIDs {
+		ruleIDs = append(ruleIDs, ruleID)
+	}
+	return ruleIDs, nil
+}
+
+func (c *multicastController) addGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleAdd,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) updateGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleUpdate,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) deleteGroupAddressForTableIDs(ruleID string, groupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		groupAddress: groupAddresses,
+		event:        ruleDelete,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) validate(iface *interfacestore.InterfaceConfig,
+	groupAddress net.IP, direction crdv1beta.Direction) (*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error) {
+
+	var ruleTypePtr *crdv1beta.NetworkPolicyType
+	action, uuid, ruleName := v1alpha1.RuleActionAllow, apitypes.UID(""), ""
+	if iface == nil {
+		// Check the if iface exists
+		klog.ErrorS(fmt.Errorf("iface should not be empty"), "")
```

Comment:
```suggestion
		// Check if the iface exists
		klog.ErrorS(nil, "iface should not be empty")
```
But I wonder if the error is really helpful here, there is no any context and there is already an error returned to client, which will eventually be logged with more context.

Reply:
removed this line